### ISSUE TITLE
MicroG is always getting disconnected when phone is in Doze (Idle) mode

### ISSUE THREAD

**Issue by wwwilq** [microg/android_packages_apps_GmsCore#1011]
MicroG is always getting disconnected when phone is in Doze (Idle) mode.

MicroG  - last version (of November 2019), excluded from battery optimization,  self-checks are OK.
In non-idle phone mode,  it works fine - all my apps get push messages timely.
Andoid - Oreo 8.1,  LOS-based ROM.

How to reproduce:

Open MicroG, open Google Cloud Messaging screen, in the upper-right corner click three-dot menu and click Advanced,  set all ping intervals = 60 seconds, return to Google Cloud Messaging screen, observe line "Current State: Connected Since ... "

Your phone must be rooted, or use adb.
Open terminal emulator, type "su", then artificially enter the phone into deep Doze (Idle) mode:
 (the commands are taken from https://developer.android.com/training/monitoring-device-state/doze-standby.html#testing_doze ):
dumpsys deviceidle force-idle  (just run the command - no need to unplug the phone from charger, no need to wait for the screen to turn off.  This command allows to quickly emulate the state of the phone it normally enters into within 1-2 hours after having been left unplugged, with screen turned-off, and motionless (deep Doze mode)).
Check if it really entered into deep doze mode:
dumpsys deviceidle   (in the end of the output there should be  mState=IDLE mLightState=OVERRIDE  for deep and light idle modes)
Now immediately switch to MicroG, open Google Cloud Messaging screen, observe line "Current State: Connected Since ... ". As this line is not auto-updated when this screen is static, you need to frequently switch from Google Cloud Messaging screen to Main screen of Microg, and then back to Google Cloud Messaging screen, and so on  - forth and back - to get updated info in the line "Current State: Connected Since ... ".  After 2-3 minutes (since entered into deep doze mode) of the switching between the microg screens forth and back, you'll finally observe "Current State: Disconnected" - and from now on it will stay disconnected forever, - until you explicitly exit doze/idle mode.  It's obvious, that while microg is in disconnected state, you won't receive any GCM push messages for your messengers.  The same disconnects in microg are observed after phone entered deep doze mode normally, on its own (within 1-2 hours after having been left unplugged, with screen turned-off, and motionless ).

To exit the artificial doze/idle mode, run:  
dumpsys deviceidle unforce ; dumpsys battery reset
Check if it really exited doze mode and returned back to normal mode:
dumpsys deviceidle  (in the end of the output there should be  mState=ACTIVE mLightState=ACTIVE  for deep and light idle modes)

What does it all mean? Probably there is some bug in microg code responsible for handling wakelocks. As I mentioned above, microg is excluded from battery optimization, so it seems that the only cause of this behavior is incorrect wakelocks handling in microg code.

How I workarounded the problem (I hope, temporary, - until the bug is fixed):  
I used command: 
dumpsys deviceidle disable
 -this command disables Doze mode totally, but only until next reboot. To make it persistent over reboots, add it into startup shell script - for example into /data/local/userinit.sh  or  (for those who have magisk): into /data/adb/service.d/any_name.sh (but never put this script into /data/adb/post-fs-data.d/ !!!).  Don't forget to add "execute" permissions for this script file, and set its owners to root/root (0/0).  
Code of the script:
        #!/system/bin/sh
        export PATH=/system/bin
        #list of running services before delay:
        cmd -l  > /data/local/cmd-l__log__before_delay.log  2>&1
        #delay in seconds - to wait until service "deviceidle" is up and running:
        sleep 60
        #list of running services after the delay:
        cmd -l  > /data/local/cmd-l__log__after_delay.log  2>&1
        dumpsys deviceidle disable  > /data/local/disable_doze_mode.log  2>&1

Reboot, and check log  /data/local/disable_doze_mode.log
Note, that without "sleep 60" before "dumpsys deviceidle disable" the script won't work correctly, because service "deviceidle" needs time to be launched by system.

**Comment by wwwilq** [comment]
At the same time, K9 email client works fine for me, I also excluded K9 from battery optimization and configured K9 to check my mailbox every 5 minutes, and it never misses the jobs, even in Doze mode.  It can be verified in BBS (Better Batter Stats) app:  If after a long period of inactivity of my phone (with disconnected charged, and lying motionless with black screen, and with Doze mode enabled in advance (dumpsys deviceidle enable))  I open BBS and select there  Screen Off -> Screen On  period  (approximately, it's a period of Doze mode), and then open Alarms tab in BBS,  I can see that number of alarms for K9 is 12 per hour (it's correct = every 5 minutes), but number of alarms for microg is far less than needed (considering my ping intervals that I configured in microg). So the problem is definitely in microg's code.  By the way, K9 is also opensource and can be found on f-droid, so probably it makes sense to look into K9's part of code, which handles wakelocks, and borrow the same approach from K9   for  microg's  wakelock handling code. It definitely needs to be improved: while using microg  I've missed already a lot of urgent notifications when my phone was in Doze mode.

**Comment by drrossum** [comment]
I can reproduce the issue you describe, but... ONLY if I set the GCM ping interval to 60 seconds.  In the log I then get messages like:
[code]
02-05 13:07:44.462  2720  2720 D GmsGcmMcsSvc: No heartbeat for 144 seconds, connection assumed to be dead after 120 seconds
02-05 13:07:44.463  2720  2720 D GmsGcmPrefs: learnTimeout: gcm_network_wifi
02-05 13:08:11.027  2720  2720 D GmsGcmMcsSvc: No heartbeat for 171 seconds, connection assumed to be dead after 120 seconds
02-05 13:08:11.027  2720  2720 D GmsGcmPrefs: learnTimeout: gcm_network_wifi
02-05 13:08:11.057  2720  2720 D GmsGcmMcsSvc: No heartbeat for 171 seconds, connection assumed to be dead after 120 seconds
02-05 13:08:11.057  2720  2720 D GmsGcmPrefs: learnTimeout: gcm_network_wifi
[/code]

If I set the GCM ping interval to 2 minutes (just long enough for the message above to not apply) then the GCM connection does not crash during deep sleep.

Interestingly, once the GCM connection crashes by setting the ping interval to 60seconds in deep sleep mode, it does not seem to recover even after ending deep sleep mode.  Only after I set the ping interval to a higher value the GCM connection comes back to life.

I wonder if this issue will affects timed-out GCM connections in general or only this particular case of 60-second timeout.

**Comment by wwwilq** [comment]
In my message #2  I outlined, that in BBS stats I observed  lack of number of "alerts /per hour" for microg during deep doze (as the opposite to K9).  This was observed many times, regardless of microg ping interval, even for ping interval=15 min.

**Comment by wwwilq** [comment]
> I can reproduce the issue you describe, but... ONLY if I set the GCM ping interval to 60 seconds.  .......  If I set the GCM ping interval to 2 minutes (just long enough for the message above to not apply) then the GCM connection does not crash during deep sleep. 

Wrong.  I easily reproduced the disconnection with 2 min intervals also.  Just needed to wait more time (double or triple time of the configured interval).   Internal microg's algorithm counts time elapsed since the last successful heartbeat (which was before entering doze mode), and after that it marks the connection as dead (after 2 o 3 missed heartbeats).  So, if your last successful heartbeat was, say at 11:00:00 AM, and with 60-sec interval you observed the disconnection state of microg, say, at 11:03, and then you immediately reconfigured ping interval to 2 min, and immediately returned back to GCM-screen of microg, then yes - you observed "connected since" message again, but NOT because the connection was actually restored (it did't happen, alas),  but just because you quickly increased the ping interval, so the internal microg's algorithm recalculated  the expected time when connection can be formally marked as dead, - basing on your new (increased) 2-min interval (instead of 60-sec one), and  due to the fact that  2(or3) * new 2-min interval = 4 (or 6) minutes have not elapsed yet (counted since that (the same) last successful heartbeat (which was before entering doze mode) at 11:00:00),  then the algorithm changed the formal connection state from "disconnected" to formal alive stated  "connected since", but the state was changed "on formal grounds" only.  With the new 2-min ping interval, you need to wait  11:00 + 4 (or 6) minutes = until 11:04 (or 11:06) and you will get disconnection again, I guarantee.

**Comment by delet-this** [comment]
I'm currently having this exact same problem on AospExtended, Android version Nougat 7.1.2. I think I didn't have this on Pie version for whatever reason.

**Comment by delet-this** [comment]
@wwwilq Looks like this is not an issue anymore with microG Services Core v0.2.11.202414

**Comment by rorlig** [comment]
@delet-this do you know what the fix was ?

**Comment by deesnook** [comment]
I can confirm this (or a similar) problem on a recent clean install of LineageOS (for microG) 17.1 with microG 0.2.16.204713 on a OnePlus 3T device. All self-checks OK, battery optimization turned off for both microG Services Core and microG Services Framework Proxy. So far I had set the ping interval to automatic. 

However, I wasn't able to reproduce the problem following the steps mentioned in the original post. 

interestingly, while the status (which changes even if I don't go back and forth from and to the main screen btw, at least the minutes counted do) was disconnected, I accidentally exited microG settings by killing all apps in the app switcher. when I opened microG again, the GCM connection was reestablished, even while still in doze mode. Is this even supposed to be possible?!?

I'll open a separate issue linking this one, too

**Comment by do1fmd** [comment]
I think I found another method to get it working. At least for a few days it seems to work properly now.

Instead of disabling idle mode on each boot, I had to put com.google.android.gsf on Idle-whitelist with this command (as root?):
`dumpsys deviceidle whitelist +com.google.android.gsf`
After reboot it will stay on the whitelist. It is not necessary to issue the above command on each reboot. It can be checked with the command "dumpsys deviceidle".

**Comment by fummymint** [comment]
> I think I found another method to get it working. At least for a few days it seems to work properly now (more testing required).
> 
> Instead of disabling idle mode on each boot, I had to put com.google.android.gsf on Idle-whitelist with this command (as root?):
> `dumpsys deviceidle whitelist +com.google.android.gsf`
> After reboot it will stay on the whitelist. It is not necessary to issue the above command on each reboot. It can be checked with the command "dumpsys deviceidle".

How is it 2 years later? you never updated us. Did this worked at all?

**Comment by do1fmd** [comment]
> > I think I found another method to get it working. At least for a few days it seems to work properly now (more testing required).
> > Instead of disabling idle mode on each boot, I had to put com.google.android.gsf on Idle-whitelist with this command (as root?):
> > `dumpsys deviceidle whitelist +com.google.android.gsf`
> > After reboot it will stay on the whitelist. It is not necessary to issue the above command on each reboot. It can be checked with the command "dumpsys deviceidle".
> 
> How is it 2 years later? you never updated us. Did this worked at all?

I use this solution for several month now and it is working.

**Comment by TheLastGimbus** [comment]
For me, installing MicroG with Magisk installer solved all problems - see #1527 :+1:

**Comment by fummymint** [comment]
> > > I think I found another method to get it working. At least for a few days it seems to work properly now (more testing required).
> > > Instead of disabling idle mode on each boot, I had to put com.google.android.gsf on Idle-whitelist with this command (as root?):
> > > `dumpsys deviceidle whitelist +com.google.android.gsf`
> > > After reboot it will stay on the whitelist. It is not necessary to issue the above command on each reboot. It can be checked with the command "dumpsys deviceidle".
> > 
> > 
> > How is it 2 years later? you never updated us. Did this worked at all?
> 
> I use this solution for several month now and it is working on my Blackview BV9900 Pro.

what's your ping interval for GCM?

**Comment by TheLastGimbus** [comment]
I've let them go on `auto` - i had too much nerves with my phone's battery dying that I don't want to assume that forcing it to `60 seconds` will help anyhow

Right now "mobile network" as well as wifi is at 30 minutes

I do admit that i do get them delayed sometimes, but not as much that it would cause me life-problems

**Comment by do1fmd** [comment]
> what's your ping interval for GCM?

How can I measure that?

**Comment by TheLastGimbus** [comment]
microG settings -> cloud messaging -> three dots -> advanced

**Comment by do1fmd** [comment]
Mobile Network and Roaming is 30 minutes, Wifi 10 minutes, other networks 5 minutes.
